### PR TITLE
#100: Fix retrieval of authentication tokens on Github enterprise hosts

### LINKED ISSUE #100: PullRequestPostAnalysisTask fail with GraphQLException: Not Found
**Describe the bug**
PullRequestPostAnalysisTask fail with GraphQLException: Not Found . This occur tying to run the pull request build and I don't see the PR decorator in Github repository. Refer exception stack trace at the bottom of the report.

**To Reproduce**
I'm running the sonar analysis through Jenkins and following is the `mvn` command use to run the analysis. 
`mvn  verify -Dsonar.pullrequest.github.repository=OrgName/RepoName -Dsonar.pullrequest.branch=origin/PR-53 -Dsonar.pullrequest.key=53 sonar:sonar -P check-snapshots,sonar -Dsonar.host.url=http://sonarqube792.com/ -Dsonar.login=****
`
I create the Github App following the instructions in the plugin README

**Expected behavior**
Should display the PR decorator in Github.

**Screenshots**
Attached sonar PR configuration screenshot - 
image

**Software Versions**
- Sonarqube Version - 7.9.1-community
- Plugin Version - sonarqube-community-branch-plugin-1.3.0.jar

**Additional context**
Exception Stack Trace
```
2020.02.25 00:40:49 ERROR ce[AXB5ygAmX2R5O-bftuvx][o.s.c.t.p.a.p.PostProjectAnalysisTasksExecutor] Execution of task class com.github.mc1arke.sonarqube.plugin.ce.pullrequest.PullRequestPostAnalysisTask failed
java.lang.IllegalStateException: Could not decorate Pull Request on Github
        at com.github.mc1arke.sonarqube.plugin.ce.pullrequest.github.GithubPullRequestDecorator.decorateQualityGateStatus(GithubPullRequestDecorator.java:38)
        at com.github.mc1arke.sonarqube.plugin.ce.pullrequest.PullRequestPostAnalysisTask.finished(PullRequestPostAnalysisTask.java:132)
        at org.sonar.ce.task.projectanalysis.api.posttask.PostProjectAnalysisTasksExecutor.executeTask(PostProjectAnalysisTasksExecutor.java:113)
        at org.sonar.ce.task.projectanalysis.api.posttask.PostProjectAnalysisTasksExecutor.finished(PostProjectAnalysisTasksExecutor.java:107)
        at org.sonar.ce.task.step.ComputationStepExecutor.executeListener(ComputationStepExecutor.java:91)
        at org.sonar.ce.task.step.ComputationStepExecutor.execute(ComputationStepExecutor.java:63)
        at org.sonar.ce.task.projectanalysis.taskprocessor.ReportTaskProcessor.process(ReportTaskProcessor.java:81)
        at org.sonar.ce.taskprocessor.CeWorkerImpl$ExecuteTask.executeTask(CeWorkerImpl.java:209)
        at org.sonar.ce.taskprocessor.CeWorkerImpl$ExecuteTask.run(CeWorkerImpl.java:191)
        at org.sonar.ce.taskprocessor.CeWorkerImpl.findAndProcessTask(CeWorkerImpl.java:158)
        at org.sonar.ce.taskprocessor.CeWorkerImpl$TrackRunningState.get(CeWorkerImpl.java:133)
        at org.sonar.ce.taskprocessor.CeWorkerImpl.call(CeWorkerImpl.java:85)
        at org.sonar.ce.taskprocessor.CeWorkerImpl.call(CeWorkerImpl.java:53)
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
        at java.base/java.util.concurrent.FutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.base/java.lang.Thread.run(Unknown Source)
Caused by: io.aexp.nodes.graphql.exceptions.GraphQLException: Not Found
        at io.aexp.nodes.graphql.Fetch.send(Fetch.java:90)
        at io.aexp.nodes.graphql.GraphQLTemplate.execute(GraphQLTemplate.java:122)
        at io.aexp.nodes.graphql.GraphQLTemplate.mutate(GraphQLTemplate.java:91)
        at com.github.mc1arke.sonarqube.plugin.ce.pullrequest.github.v4.GraphqlCheckRunProvider.createCheckRun(GraphqlCheckRunProvider.java:162)
        at com.github.mc1arke.sonarqube.plugin.ce.pullrequest.github.GithubPullRequestDecorator.decorateQualityGateStatus(GithubPullRequestDecorator.java:36)
        ... 19 common frames omitted
```

### PR DESCRIPTION
**Pull request by kennyko**
Updated configuration description to be in line with the documentation on https://docs.sonarqube.org/latest/instance-administration/github-application/
Add support for GitHub Enterprise installations with both a `/v3` in the URL and without.

Resolves #100

### CODE REVIEW

File: src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/github/v3/RestApplicationAuthenticationProvider.java [hunk 1]
```diff
@@ -131,6 +131,13 @@ public RepositoryAuthenticationToken getInstallationToken(String apiUrl, String
                 "No token could be found with access to the requested repository with the given application ID and key");
     }
 
+    private String getV3Url(String apiUrl) {
+        if (apiUrl.matches("^.*/api(/v3)?/?$")) {
+            return apiUrl.replaceFirst("/api(/v3)?/?", "/api/v3");
+        } else {
+            return apiUrl.replaceFirst("/$", "");
+        }
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/github/v3/RestApplicationAuthenticationProvider.java]: I'm not a fan of the regular expressions. Could we do this in 2 steps:

```suggestion
        if (apiUrl.endWith("/") {
            apiUrl = apiUrl.substring(0, apiUrl.length() - 1);
        }
        
        if (apiUrl.endsWith("/api/v3")) {
            apiUrl = apiUrl.substring(0, apiUrl.lengh() - 7);
        }
        
        return apiUrl;
```

File: src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/github/v4/GraphqlCheckRunProvider.java [hunk 2]
```diff
@@ -174,6 +174,14 @@ public void createCheckRun(AnalysisDetails analysisDetails, UnifyConfiguration u
         }
     }
 
+    private String getGraphqlUrl(String apiUrl) {
+        if (apiUrl.matches("^.*/api(/v3)?/?$")) {
+            return apiUrl.replaceFirst("/api(/v3)?/?", "/api/graphql");
+        } else {
+            return apiUrl.replaceFirst("/?(graphql)?/?$", "/graphql");
+        }
+    }
```
Comment (mc1arke) [inline src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/github/v4/GraphqlCheckRunProvider.java]: Similarly, can we do this without regex?

File: src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/github/v3/RestApplicationAuthenticationProvider.java (hunk 1 above)
Comment (kennyko) [reply src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/github/v3/RestApplicationAuthenticationProvider.java]: Sure, will do.

**Comment by kennyko** [comment]
@mc1arke are you able to take a look at this pull request again?

**Comment by mc1arke** [comment]
Please squash your commits so we have a single commit for this feature

**Comment by kennyko** [comment]
Done

**Comment by mc1arke** [comment]
Merged. Thanks for the contribution!

**Pull request merged**